components/cdc/src/endpoint.rs

```diff
@@ -120,25 +142,331 @@ impl Endpoint {
+        let checkpoint_ts = self.checkpoint_ts;
+        let build_resolver = self.build_resolver;
+        info!("async incremental scan";
+            "region_id" => region.get_id(),
+            "downstream_id" => ?downstream_id);
+
+        // spawn the task to a thread pool.
+        let region_id = region.get_id();
+        self.workers
+            .spawn(lazy(move || {
+                let mut resolver = if build_resolver {
+                    Some(Resolver::new())
+                } else {
+                    None
+                };
+
+                // Time range: (checkpoint_ts, current]
+                let current = TimeStamp::max();
+                let mut scanner = ScannerBuilder::new(snap, current, false)
+                    .range(None, None)
+                    .build_delta_scanner(checkpoint_ts)
+                    .unwrap();
+                let mut done = false;
+                while !done {
+                    let entries =
+                        match Self::scan_batch(&mut scanner, batch_size, resolver.as_mut()) {
+                            Ok(res) => res,
+                            Err(e) => {
+                                error!("cdc scan entries failed"; "error" => ?e);
+                                // TODO: record in metrics.
+                                if let Err(e) = sched.schedule(Task::Deregister {
+                                    region_id,
+                                    downstream_id: Some(downstream_id),
+                                    err: Some(e),
+                                }) {
+                                    error!("schedule task failed"; "error" => ?e);
+                                }
+                                return Ok(());
+                            }
+                        };
+                    // If the last element is None, it means scanning is finished.
+                    if entries.is_empty() {
+                        done = true;
+                    }
+                    debug!("cdc scan entries"; "len" => entries.len());
+                    let scanned = Task::IncrementalScan {
+                        region_id,
+                        downstream_id,
+                        entries,
+                    };
+                    if let Err(e) = sched.schedule(scanned) {
+                        error!("schedule task failed"; "error" => ?e);
+                        return Ok(());
+                    }
+                }
+
+                if let Some(resolver) = resolver {
+                    Self::finish_building_resolver(resolver, region, sched);
+                }
+
+                Ok(())
+            }))
+            .unwrap();
+    }
+
+    fn scan_batch<S: Snapshot>(
+        scanner: &mut DeltaScanner<S>,
+        batch_size: usize,
+        resolver: Option<&mut Resolver>,
+    ) -> Result<EntryBatch> {
+        let mut entry_batch = EntryBatch::with_capacity(batch_size);
```

Comment:
It allocates every time we call this function, could you reuse the buffer?

Reply:
oh, good idea

Reply:
The batch is then sent back to the scheduler at line 252 which makes it difficult to reuse in the current architecture.